chore: move embark site into mono repo

PascalPrecht opened the pull request:
@michaelsbradleyjr please double check if there's adjustments needed in `lerna.json`.

Right now, I think the most pragmatic step forward is to have lerna ideally ignore `./site`.

PascalPrecht commented on deploy-site.sh:
```diff
@@ -0,0 +1,32 @@
+git subtree split -P site/public -b deploy-dist
+
+# Push source and build repos.
+git push -f embark-site deploy-dist:gh-pages
```
**IMPORTANT** the origin is called `embark-site`, so everyone who wants to deploy needs to make sure to have it set up accordingly

michaelsbradleyjr replied:
Can we use a git command to check whether the `embark-site` remote is setup in the local repository?

Can we employ some logic similar to that of `scripts/release.js` re: `--release-branch`, pertaining to branches of `embark-framework/embark`? One, check that the current branch is a match for `--release-branch`, which defaults to `master` if not specified in the cli. Two, check that the local and remote branches are in sync.

If we make it so this script is invoked by `scripts/release.js` then it's not an issue. But I'm thinking that's probably not always the case. For example, fixing typos or grammatical errors, or adding a post — such changes won't always coincide with an embark release. So it seems we may be invoking this script manually sometimes and automatically other times.

PascalPrecht replied:
> One, check that the current branch is a match for --release-branch, which defaults to master if not specified in the cli.

Since the release branch will always be `gh-pages` there's probably no point in checking if the current branch is a match for sth. like `--release-branch`. So even if the default is `master`, the website will be pushed to `gh-pages` on the embark-site origin. Unless we have a reason to make this configurable, but I don't see any at this point.

> Two, check that the local and remote branches are in sync.

First I thought: Yes we can do that. But then I realized, what if I really just want to put out some typo fixes and have to do that from a different branch (maybe prior to a feature that has landed incl. docs but should be released yet). In such case, there is no remote branch that can be in sync with the local one.

PascalPrecht replied:
> If we make it so this script is invoked by scripts/release.js then it's not an issue. But I'm thinking that's probably not always the case. For example, fixing typos or grammatical errors, or adding a post — such changes won't always coincide with an embark release. So it seems we may be invoking this script manually sometimes and automatically other times.

Exactly. We definitely want to keep this separate. Embark and Website should be releasable without releasing the other.

michaelsbradleyjr replied:
> Since the release branch will always be `gh-pages` there's probably no point...

I meant with respect to local branch and the corresponding remote branch on github `embark-framework/embark`, not `embark-framework/embark-site`.

If I'm locally on `master` then before deploying to remote `gh-pages` branch on `embark-framework/embark-site` is it important to check that remote `master` branch on `embark-framework/embark` matches local `master`?

michaelsbradleyjr replied:
> But then I realized, what if I really just want to put out some typo fixes... there is no remote branch that can be in sync with the local one.

Shouldn't the typo fixes land in `master` of `embark-framework/embark` before the deployment to `gh-pages` of `embark-framework/embark-site`?

PascalPrecht replied:
Absolutely. But imagine this scenario (which is very close to what we've discussed with regards to our release flow):

- Typo fixes landed in master
- New features landed in master in the meantime, including their docs
- typo fixes need to be deployed, but we don't wanna put the feature docs out yet because Embark hasn't made a new release yet and is not planning to do so right now

^ For this I would probably just go back in time, check out that commit that has the typo fixes, or even make a branch and cherry-pick several fixes if needed and run a deploy from there.

This is pretty much the same thing we've done for `4.0.x` releases for Embark, just with the crucial difference that we don't keep the commit history for the released website.

michaelsbradleyjr replied:
> This is pretty much the same thing we've done for `4.0.x` releases for Embark, just with the crucial difference that we don't keep the commit history for the released website.

Thanks, the explanation makes sense.

> typo fixes need to be deployed, but we don't wanna put the feature docs out yet because Embark hasn't made a new release yet and is not planning to do so right now

Would it make sense if deployed as nightlies of the website, corresponding to nightlies of embark?

PascalPrecht commented on deploy-site.sh:
```diff
@@ -0,0 +1,32 @@
+git checkout master
+git branch -D deploy
+
+
```
I had to barely touch this script. Only thing was to move it to the root of our repo, otherwise we can't use `git subtree`.

michaelsbradleyjr replied:
Can you explain?  Does it have to do with the current working directory and how the script is invoked?

PascalPrecht replied:
`git subtree` can only be invoked from the root of a git repository.

michaelsbradleyjr replied:
What if the script is in `<root>/scripts/deploy-site.sh`, same `scripts/` dir where `release.js` is located?

Will it work to invoke it this way?

```
$ cd <root>
$ scripts/deploy-site.sh
```

The current working directory would be the root of the git repository even though the script is located in `scripts/`.

PascalPrecht replied:
This might work! Lemme try that.

PascalPrecht replied:
I'm sorry actually, the reason I moved the deploy-site.sh script in the root was because exactly that didn't work haha

michaelsbradleyjr replied:
No problem, I wasn't sure if maybe you were invoking the script from the directory in which it was located instead of invoking it from the root.

PascalPrecht replied:
I stand corrected. @michaelsbradleyjr your guess was spot on. I was testing a different setting

PascalPrecht commented on site/LICENSE:
```diff
@@ -0,0 +1,7 @@
+
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WI ...
```
This one is most likely from hexo's original fork

PascalPrecht commented on site/gulpfile.js:
```diff
@@ -0,0 +1,115 @@
+'use strict';
```
Putting a comment here because I can't in `./site/debug.log` ... `debug.log needs to be ignored and thrown out.

PascalPrecht commented on site/package.json:
```diff
@@ -0,0 +1,49 @@
+{
+  "name": "hexo-site",
+  "version": "0.0.0",
```
We probably want to change these to `@embark/site` etc

jrainville replied:
For sure 👍

michaelsbradleyjr replied:
Even though it's a private package, maybe it's best to not use a namespace we don't own. Maybe `embark-site` instead?

PascalPrecht replied:
Yeap!

jrainville commented on .gitignore:
```diff
@@ -18,3 +18,8 @@ yarn-debug.log*
+site/node_modules
+site/public
+site/db.json
+site/packge-log.json
```
Can't those just live inside the site folder? Inside another `.gitignore`

michaelsbradleyjr replied:
Agreed, and a couple of points:

There's no need to put `node_modules` in `site/.gitignore` because it's covered by the root `.gitignore`.

Also, I think there's a typo in `site/packge-log.json`, I'm guessing it's supposed to be `site/package-lock.json`. That one is also unnecessary because it's covered by root `.gitignore`.

So `site/.gitignore` should have:

```
public
db.json
!/yarn.lock
```

The `!/yarn.lock` is so that `yarn.lock` of the site is not ignored, since we're not managing the site as part of the workspace.... though I'm starting to wonder if it should be part of the workspace.

PascalPrecht replied:
Will update!

PascalPrecht commented:
> I guess you put the site not inside packages/ because otherwise it would be deployed to npm or something?

Yes, at least we want to clearly differentiate between embark packages and the website. Probably easiest for now if lerna completely ignores `./site`, waiting for @michaelsbradleyjr input on that.

michaelsbradleyjr commented on deploy-site.sh:
```diff
@@ -0,0 +1,32 @@
+cd site && npx hexo generate
+
+# delete old deploy branch
+git branch -D deploy
```
I'm wondering if it might be safer to "scope" the branch names: `embark-site-deploy` and `embark-site-deploy-dist`.  I don't think `deploy` and `deploy-dist` will conflict with anyone's local branches, but maybe better to play it safe.

PascalPrecht replied:
Good point!

PascalPrecht commented:
@jrainville @michaelsbradleyjr addressed your comments:

- Optimized some of the illustration assets which were couple MB. Please note that we still have big assets in this project. This is because we're using animated gifs for the documentation
- Rewrote the deploy script in Node, added a check that the correct remote is configured
- Added an `deploy:site` npm script to the root's package.json
- Removed a lot of files within the site project that we don't use any more (patternlab stuff etc)
- Re-added site/.gitignore as discussed.

michaelsbradleyjr commented on site/deploy-site.js:
```diff
@@ -0,0 +1,46 @@
+const execWithOutput = (cmd) => execSync(cmd, { stdio: 'inherit' });
+
+const DEPLOY_REPOSITORY = 'https://github.com/embark-framework/embark-site';
+const DEPLOY_ORIGIN = 'embark-site';
```
Very small thing.  If I understand correctly (maybe not), the term *origin* is more of a convention for the primary remote repository.  The general term is *remote*. See, for example, the OPTIONS section of `man git-push`:

```
OPTIONS
       <repository>
           The "remote" repository that is destination of a push operation. This parameter can be either a
           URL (see the section GIT URLS below) or the name of a remote (see the section REMOTES below).
```

See also: https://github.com/embark-framework/embark/blob/master/scripts/release.js#L21-L22

There could be a good reason to set a default `DEPLOY_REMOTE`, with possibility of override (see link above), and that's the ability to do test releases against a "fake" remote, i.e. a remote that points to another repository on the local filesystem. That's the approach I use (together with verdaccio for the registry aspect) to do local test releases of embark.

PascalPrecht replied:
> The general term is remote. 

This is correct. And semantics are important. Will change this to `DEPLOY_REMOTE`.

> There could be a good reason to set a default DEPLOY_REMOTE, with possibility of override (see link above), and that's the ability to do test releases against a "fake" remote, i.e. a remote that points to another repository on the local filesystem. That's the approach I use (together with verdaccio for the registry aspect) to do local test releases of embark. DEPLOY_REPOSITORY also wouldn't be needed — it's up to the user to make sure the remote points to the correct place, which doesn't seem too unreasonable.

Yes we can make this configurable.

michaelsbradleyjr commented on site/deploy-site.js:
```diff
@@ -0,0 +1,46 @@
+
+const SITE_DIR = 'site';
+const PUBLIC_DIR = 'public';
+const ROOT_DIR = process.cwd();
```
```suggestion
const ROOT_DIR = path.join(__dirname, '..');
```

Could be a more reliable way to calculate the root, so that it's correct regardless of the directory from which the script was invoked.

PascalPrecht replied:
Ah thanks!

michaelsbradleyjr commented:
@PascalPrecht @jrainville What if we setup a repo like `embark-site-assets` and could make its default branch `gh-pages`. Images and other bulky static things could be stored there and then we link to those assets from the site.

I don't want to overly complicate things, but hoping we can somehow avoid repo bloat with static assets.

PascalPrecht commented:
> What if we setup a repo like embark-site-assets and made its default branch gh-pages. Images and other bulky static things could be stored there and then we link to those assets from the site.

Certainly an option. Just comes with the side effect that, when running locally (and offline) those assets are missing unless the assets repo is setup and served locally as well.

This may or may not be a problem.

Is this something we want to decide before this PR gets merged? Or can we iterate and do that in another step

jrainville commented:
> when running locally (and offline) those assets are missing unless the assets repo is setup and served locally as well.

jrainville commented:
> when running locally (and offline) those assets are missing unless the assets repo is setup and served locally as well.

They would be available if we always use the remote url (eg `embark.status.im/images/something.gif`)

However, there still comes the problem that adding assets to the website is still quite a pain, because you need to add them to the other repo and publish.

michaelsbradleyjr commented on site/deploy-site.js:
```diff
@@ -0,0 +1,52 @@
+const ROOT_DIR = path.join(__dirname, '..');
+
+const remote = args['deploy-remote'] || DEPLOY_REMOTE;
+const repository = args['deploy-repository'] || DEPLOY_REPOSITORY;
```
It's awesome to have this be configurable, great work!

But I don't understand the strict need for `DEPLOY_REPOSITORY`. Imagine I want to use `my-fake-embark-site` remote. When I setup that remote, I will have pointed it to something like `/Users/michael/temp/my-fake-embark-site`, i.e. a path to another repo on my local filesystem.

So then for `deploy-site` script to work, I need to supply that same URL on the command-line, and in practice I'll either recall it from memory or copy/paste from my git config.

What if we checked whether `remote === DEPLOY_REMOTE`, i.e. it's `embark-site` (the default), and only in that case check that remote's URL `=== DEPLOY_REPOSITORY`?

That would catch a badly configured `embark-site` remote for real deployments and not be a hassle when doing test deployments with `--deploy-remote`.

PascalPrecht replied:
Sorry, you're right, this setting doesn't make a lot of sense anymore when making those things configurable via passing down a remote. Will update.

michaelsbradleyjr commented on site/deploy-site.js:
```diff
@@ -0,0 +1,52 @@
+const remoteDeployBranch = args['deploy-branch'] || REMOTE_DEPLOY_BRANCH;
+
+function main() {
+  const hasCorrectOrigin = execSync(`git remote -v | grep ${remote}`).toString().indexOf(remote) > -1;
```
```suggestion
  const hasCorrectRemote = execSync(`git remote -v | grep ${remote}`).toString().indexOf(remote) > -1;
```

michaelsbradleyjr commented on site/deploy-site.js:
```diff
@@ -0,0 +1,52 @@
+
+function main() {
+  const hasCorrectOrigin = execSync(`git remote -v | grep ${remote}`).toString().indexOf(remote) > -1;
+  const hasCorrectOriginRepo = execSync(`git remote -v | grep ${remote}`).toString().indexOf(repository) > -1;
```
```suggestion
  const hasCorrectRemoteRepo = remote !== DEPLOY_REMOTE ? true : execSync(`git remote -v | grep ${remote}`).toString().indexOf(repository) > -1;
```

See: https://github.com/embark-framework/embark/pull/1522/files#r278630603.

michaelsbradleyjr commented on site/deploy-site.js:
```diff
@@ -0,0 +1,52 @@
+  const hasCorrectOrigin = execSync(`git remote -v | grep ${remote}`).toString().indexOf(remote) > -1;
+  const hasCorrectOriginRepo = execSync(`git remote -v | grep ${remote}`).toString().indexOf(repository) > -1;
+
+  if (!hasCorrectOrigin || !hasCorrectOriginRepo) {
```
```suggestion
  if (!hasCorrectRemote || !hasCorrectRemoteRepo) {
```

PascalPrecht commented:
Thanks for the suggestions @michaelsbradleyjr ! Updated accordingly :)

michaelsbradleyjr commented on site/deploy-site.js:
```diff
@@ -0,0 +1,51 @@
+
+function main() {
+  const hasCorrectRemote = execSync(`git remote -v | grep ${remote}`).toString().indexOf(remote) > -1;
+  const hasCorrectRemoteRepo = remote !== DEPLOY_REMOTE ? true : execSync(`git remote -v | grep ${remote}`).toString().indexOf(repository) > -1;
```
```suggestion
  const hasCorrectRemoteRepo = remote !== DEPLOY_REMOTE ? true : execSync(`git remote -v | grep ${remote}`).toString().indexOf(DEPLOY_REPOSITORY) > -1;
```

michaelsbradleyjr commented on site/deploy-site.js:
```diff
@@ -0,0 +1,51 @@
+  const hasCorrectRemote = execSync(`git remote -v | grep ${remote}`).toString().indexOf(remote) > -1;
+  const hasCorrectRemoteRepo = remote !== DEPLOY_REMOTE ? true : execSync(`git remote -v | grep ${remote}`).toString().indexOf(DEPLOY_REPOSITORY) > -1;
+
+  if (!hasCorrectOrigin || !hasCorrectRemoteRepo) {
```
```suggestion
  if (!hasCorrectRemote || !hasCorrectRemoteRepo) {
```
